src/librustc_mir/const_eval.rs

```diff
@@ -465,13 +465,16 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
             }
         }
 
+        // Emit a warning if this is the first time we have triggered the loop detector. This means
+        // we have been executing for quite some time.
         let span = ecx.frame().span;
-        ecx.machine.loop_detector.observe_and_analyze(
-            *ecx.tcx,
-            span,
-            &ecx.memory,
-            &ecx.stack[..],
-        )
+        if ecx.machine.loop_detector.is_empty() {
+            // FIXME(#49980): make this warning a lint
+            ecx.tcx.sess.span_warn(span,
+                "Constant evaluating a complex constant, this might take some time");
+        }
```

Comment:
Why did you move this out? Seems like a layering violation for this code to care about "emptiness" of the loop detector.

Reply:
Oh I see, you could remove the `tcx` and `span` arguments then. That is a good argument, but `is_empty` is not a great name.

Reply:
`has_been_invoked`?

Reply:
No that's negated... `is_unused`?

Reply:
`!has_been_invoked` is negated but `!is_unused` isn't?

Reply:
> !has_been_invoked is negated but is_unused isn't?

Well I wanted to avoid the `!` is all. It's way too easy to miss.